File: arrays_and_hashing/valid_anagram.py
# Q:242 - https://leetcode.com/problems/valid-anagram/

class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        
        sMap = self.__createMap(s)
        tMap = self.__createMap(t)

        for ch in tMap:
            if ch not in sMap or tMap[ch] != sMap[ch]:
                return False
        
        return True
    
    # creates a map of the string character against their frequency of occurence
    def __createMap(self, s: str) -> dict:
        map = dict()

        for ch in s:
            map[ch] = map.get(ch, 0) + 1

        return map


c = Solution()

s = "anagram"
t = "nagaram"

s1 = "rat"
t1 = "car"

print(c.isAnagram(s, t))
print(c.isAnagram(s1, t1))

File: arrays_and_hashing/valid_anagram.js
// Q:242 - https://leetcode.com/problems/valid-anagram/

/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isAnagram = function (s, t) {
  if (s.length !== t.length) {
    return false;
  }

  const sMap = {},
    tMap = {};

  for (let i = 0; i < s.length; i++) {
    const currentChar = s[i];
    if (!sMap.hasOwnProperty(currentChar)) {
      sMap[currentChar] = 0;
    }

    sMap[currentChar] += 1;
  }

  for (let i = 0; i < t.length; i++) {
    const currentChar = t[i];
    if (!tMap.hasOwnProperty(currentChar)) {
      tMap[currentChar] = 0;
    }

    tMap[currentChar] += 1;
  }

  for (let key in tMap) {
    if (!sMap.hasOwnProperty(key) || tMap[key] !== sMap[key]) {
      return false;
    }
  }

  return true;
};
